libsolidity/analysis/ConstantEvaluator.cpp

```diff
@@ -407,3 +443,37 @@ void ConstantEvaluator::endVisit(TupleExpression const& _tuple)
 	if (!_tuple.isInlineArray() && _tuple.components().size() == 1)
 		m_values[&_tuple] = evaluate(*_tuple.components().front());
 }
+
+void ConstantEvaluator::endVisit(FunctionCall const& _functionCall)
+{
+	using util::keccak256;
+	using util::h256;
+
+	auto const* builtinFunction = dynamic_cast<MagicVariableDeclaration const*>(ASTNode::referencedDeclaration(_functionCall.expression()));
+	if (!builtinFunction)
+		return;
+
+	auto const* functionType = builtinFunction->functionType(true);
+	solAssert(functionType);
+	switch (functionType->kind())
+	{
+		case FunctionType::Kind::ERC7201:
+		{
+			solAssert(_functionCall.arguments().size() == 1);
+			auto stringArg = evaluate(*(_functionCall.arguments()[0].get()));
+			if (!stringArg || !std::holds_alternative<std::string>(stringArg->value))
+				return;
+
+			auto bytesValue = keccak256(std::get<std::string>(stringArg->value));
+			bytesValue.data()[31] -= 1;
```

Comment:
I don't think this matches the intended expression. When the last byte is 0, it will flip to `0xff` but the rest of the number will remain unchanged.

You should instead treat the whole expression as a number and subtract one from it.

Reply:
Please add a test case which in which the hash of the string has zero in its last byte.